ksql-common/src/main/java/io/confluent/ksql/function/KsqlTableFunction.java

```diff
@@ -15,21 +15,40 @@
 
 package io.confluent.ksql.function;
 
-import io.confluent.ksql.schema.ksql.types.SqlType;
+import com.google.errorprone.annotations.Immutable;
+import io.confluent.ksql.name.FunctionName;
 import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
 import org.apache.kafka.connect.data.Schema;
 
 /**
  * A wrapper around the actual table function which provides methods to get return type and
  * description, and allows the function to be invoked.
  */
-public interface KsqlTableFunction extends FunctionSignature {
+@Immutable
+public class KsqlTableFunction extends KsqlFunctionBase {
 
-  Schema getReturnType();
+  private final FunctionInvoker invoker;
+  private final Object udtf;
 
-  SqlType returnType();
+  public KsqlTableFunction(
+      final Function<List<Schema>, Schema> returnSchemaProvider,
+      final FunctionName functionName,
+      final Schema outputType,
+      final List<Schema> arguments,
+      final String description,
+      final FunctionInvoker functionInvoker,
+      final Object udtf
+  ) {
+    super(returnSchemaProvider, outputType, arguments, functionName, description,
+        "", false
+    );
+    this.invoker = Objects.requireNonNull(functionInvoker, "functionInvoker");
+    this.udtf = Objects.requireNonNull(udtf, "udtf");
+  }
 
-  List<?> apply(Object... args);
-
-  String getDescription();
+  public List<?> apply(final Object... args) {
```

Comment:
I agree. The way that KsqlFunction gets invoked seems overelaborate. I'll see if I can refactor it to be similar to KsqlTableFunction

Reply:
I would like to make this change and the previous one in a follow up PR in the stack if that's OK? Reason being there might be a merge nightmare otherwise as an existing PR later in this stack does another refactoring in this area.

Reply:
I dug into this a bit. The reason that, for, UDFs, newInstance is called, and the invocation is made on the Kudf instance seems to be because UDFs can have ksql properties overridden on a per query basis. There only seems to be one UDF that actually does this - SUBSTRING: https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java#L41
And it's configured in the factory here: https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java#L169
For this to work, the system needs to ensure that each function has it's own object instance, because using this configure trick means the UDFs aren't stateless. 
IMHO this is a really ugly hack, and it should be done a different way, but that needs more thought.

Reply:
ack. makes sense